core/frontend/src/GeoServices.ts

```diff
@@ -3,240 +3,287 @@
 // cspell:ignore GCRS
-import { XYZProps } from "@itwin/core-geometry";
 import {
-  GeoCoordinatesRequestProps, GeoCoordinatesResponseProps, GeoCoordStatus, GeographicCRSProps, IModelCoordinatesRequestProps, IModelCoordinatesResponseProps,
-  IModelReadRpcInterface, PointWithStatus,
+  assert, BeEvent, Dictionary, Logger, SortedArray,
+} from "@itwin/core-bentley";
+import { WritableXYAndZ, XYAndZ, XYZProps } from "@itwin/core-geometry";
+import {
+  GeoCoordinatesResponseProps, GeoCoordStatus, GeographicCRSProps, IModelCoordinatesResponseProps, IModelReadRpcInterface, PointWithStatus,
 } from "@itwin/core-common";
 import { IModelConnection } from "./IModelConnection";
+import { FrontendLoggerCategory } from "./FrontendLoggerCategory";
 
-/** Response to a request to obtain imodel coordinates from cache.
- * @internal
+/** Options used to create a [[CoordinateConverter]].
+ * @internal exported strictly for tests.
  */
-export interface CachedIModelCoordinatesResponseProps {
-  /** An array of the same length as the input array, with undefined entries at indices corresponding to points not found in cache. */
-  result: Array<PointWithStatus | undefined>;
-  /** An array of points in the input array which were not found in the cache, or undefined if all points were found in the cache. */
-  missing?: XYZProps[];
+export interface CoordinateConverterOptions {
+  /** The iModel for which to perform the conversions. */
+  iModel: IModelConnection;
+  /** Asynchronously convert each point. The resultant array should have the same number and order of points as the input. */
+  requestPoints: (points: XYAndZ[]) => Promise<PointWithStatus[]>;
+  /** Maximum number of points to include in each request. Default: 300. */
+  maxPointsPerRequest?: number;
 }
 
-// this class is used to cache results from conversion of geoCoordinates to IModelCoordinates.
-class GCtoIMCResultCache {
-  // see fast-memoize npm package where the author demonstrated that an object is the fastest
-  //     lookup (faster than Map), and JSON.stringify is the fastest serializer.
-  private _cache: any;
-  private _iModel: IModelConnection;
-  private _source: string;
-
-  constructor(iModel: IModelConnection, source: string) {
-    this._iModel = iModel;
-    this._cache = {};
-    this._source = source;
-  }
+function compareXYAndZ(lhs: XYAndZ, rhs: XYAndZ): number {
+  return lhs.x - rhs.x || lhs.y - rhs.y || lhs.z - rhs.z;
+}
 
-  /** @internal */
-  public findInCache(geoPoints: XYZProps[]): CachedIModelCoordinatesResponseProps {
-    const result: Array<PointWithStatus | undefined> = [];
-    let missing: XYZProps[] | undefined;
-    for (const geoPoint of geoPoints) {
-      const key = JSON.stringify(geoPoint);
-      const imodelCoord = this._cache[key];
-      result.push(imodelCoord);
-      if (undefined === imodelCoord) {
-        if (undefined === missing)
-          missing = [];
+function cloneXYAndZ(xyz: XYAndZ): XYAndZ {
```

Comment:
do we really need to add this? couldnt we just use the spread operator?

Reply:
See comment in description about `XYZProps` is a `PITA`. It can be an array or an object {x, y, z }. Any of the array elements or the object properties can be undefined.

Reply:
Since its an internal/local function it's ok and no objections to that; i see you're just passing it into cloneFunction of the Arrays so nbd there